### PR TITLE
Implement RFC: Improved 32-bit LNGLAT projection mode

### PR DESCRIPTION
**Pull request by georgios-uber**
COORDINATE_SYSTEM_EXPERIMENTAL_LNG_LAT

For #1937 

#### Background
See RFC.

#### Change List
- Add new projection mode

### CODE REVIEW

File: modules/core/src/shaderlib/project/project.glsl.js [hunk 1]
```diff
@@ -38,6 +39,7 @@ uniform vec2 project_uViewportSize;
 uniform float project_uDevicePixelRatio;
 uniform float project_uFocalDistance;
 uniform vec3 project_uCameraPosition;
+uniform vec2 project_coordinate_origin;
```
[the remark this answers is not in the document]
Comment (georgios-uber) [reply modules/core/src/shaderlib/project/project.glsl.js]: Not sure what project_uCenter has. Let me find out.

File: modules/core/src/shaderlib/project/project.glsl.js [hunk 2]
```diff
@@ -95,7 +97,7 @@ vec2 project_mercator_(vec2 lnglat) {
 //
 // Projects lnglats (or meter offsets, depending on mode) to pixels
 //
-vec4 project_position(vec4 position) {
+vec4 project_position(vec4 position, vec2 positionLOW) {
```
[the remark this answers is not in the document]
Comment (georgios-uber) [reply modules/core/src/shaderlib/project/project.glsl.js]: OK, I also need to fix all project_position() variants.

File: modules/core/src/shaderlib/project/project.glsl.js [hunk 3]
```diff
@@ -105,6 +107,12 @@ vec4 project_position(vec4 position) {
     );
   }
 
+  if (project_uCoordinateSystem == COORDINATE_SYSTEM_EXPERIMENTAL_LNG_LAT) {
+    float X = position.x - project_coordinate_origin.x;
```
[the remark this answers is not in the document]
Comment (georgios-uber) [reply modules/core/src/shaderlib/project/project.glsl.js]: I will make this good/proper code.

File: modules/core/src/shaderlib/project/project.glsl.js [hunk 4]
```diff
@@ -115,6 +123,15 @@ vec4 project_position(vec4 position) {
   return project_offset_(position_modelspace);
 }
 
+vec4 project_position(vec4 position) {
+  return project_position(position, vec2(0.0, 0.0));
+}
+
+vec3 project_position(vec3 position, vec2 positionLOW) {
```
[the remark this answers is not in the document]
Comment (georgios-uber) [reply modules/core/src/shaderlib/project/project.glsl.js]: Yes, I need to fix all variants and make sure they also work in all modes.

File: modules/core/src/shaderlib/project/viewport-uniforms.js [hunk 5]
```diff
@@ -54,8 +55,20 @@ function calculateMatrixAndOffset({
   let {viewProjectionMatrix} = viewport;
 
   let projectionCenter;
+  let usedCoordinateSystem = coordinateSystem;
```
[the remark this answers is not in the document]
Comment (georgios-uber) [reply modules/core/src/shaderlib/project/viewport-uniforms.js]: OK, was trying to find better name.

File: modules/core/src/shaderlib/project/viewport-uniforms.js [hunk 6]
```diff
@@ -54,8 +55,20 @@ function calculateMatrixAndOffset({
...
+  let usedCoordinateOrigin = coordinateOrigin;
+
+  if (coordinateSystem === COORDINATE_SYSTEM.EXPERIMENTAL_LNG_LAT) {
+    if (coordinateZoom < EXPERIMENTAL_LNG_LAT_ZOOM_THRESHOLD) {
+      usedCoordinateSystem = COORDINATE_SYSTEM.LNGLAT;
```
[the remark this answers is not in the document]
Comment (georgios-uber) [reply modules/core/src/shaderlib/project/viewport-uniforms.js]: Agree.

File: modules/core/src/shaderlib/project/viewport-uniforms.js [hunk 7]
```diff
@@ -54,8 +55,20 @@ function calculateMatrixAndOffset({
...
+    } else {
+      const x = Math.fround(viewport.longitude);
+      const y = Math.fround(viewport.latitude);
+      usedCoordinateOrigin = [x, y];
```
[the remark this answers is not in the document]
Comment (georgios-uber) [reply modules/core/src/shaderlib/project/viewport-uniforms.js]: I think so, but let me look more.

File: modules/layers/src/line-layer/line-layer.js [hunk 8]
```diff
@@ -130,13 +130,13 @@ export default class LineLayer extends Layer {
   }
 
   calculateInstanceSourceTargetPositions64xyLow(attribute) {
-    const isFP64 = this.is64bitEnabled();
-    attribute.isGeneric = !isFP64;
```
[the remark this answers is not in the document]
Comment (georgios-uber) [reply modules/layers/src/line-layer/line-layer.js]: Yes, but we need to update all the layers 😟😟😟😟

[the remark this answers is not in the document]
Comment (Pessimistress) [reply modules/layers/src/line-layer/line-layer.js]: Why can't we just change `Layer.is64bitEnabled()`?

File: modules/core/src/shaderlib/project/viewport-uniforms.js [hunk 9]
```diff
@@ -34,6 +34,7 @@ const VECTOR_TO_POINT_MATRIX = [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0];
 const IDENTITY_MATRIX = [1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1];
 const DEFAULT_PIXELS_PER_UNIT2 = [0, 0, 0];
 const DEFAULT_COORDINATE_ORIGIN = [0, 0, 0];
+const EXPERIMENTAL_LNG_LAT_ZOOM_THRESHOLD = 12;
```
Comment (Pessimistress) [inline modules/core/src/shaderlib/project/viewport-uniforms.js]: Thoughts about the magic threshold:
- We need some comments on how this number is determined. At zoom level 12, viewport height covers `180 / Math.pow(2, 12) ~= 0.0439 (degrees)`. According to the tests [here](https://github.com/uber-common/viewport-mercator-project/blob/master/docs/articles/offset-accuracy.md) the lngLat offset mode can handle at least another zoom level with errors less than 0.1 pixels.
- We can easily add some test cases to compare projection results right below and above this threshold. The goal is that during smooth zoom in/out there is no visible "jump" for the same coordinate.

**Comment by Pessimistress** [comment]
@georgios-uber Link to the RFC and issue in the PR summary please

File: modules/core/src/shaderlib/project/viewport-uniforms.js (hunk 9 above)
Comment (georgios-uber) [reply modules/core/src/shaderlib/project/viewport-uniforms.js]: There is currently no visible jump.
I think we can probably go a little bit more like maybe 14.

File: modules/layers/src/line-layer/line-layer.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (georgios-uber) [reply modules/layers/src/line-layer/line-layer.js]: https://github.com/uber/deck.gl/pull/1963

File: modules/core/src/shaderlib/project/viewport-uniforms.js (hunk 9 above)
Comment (Pessimistress) [reply modules/core/src/shaderlib/project/viewport-uniforms.js]: My point being, this logic should be captured by some form of documentation or test case. Both accuracy and discontinuity can be measured quantitively. Arbitrary manual test case does not support the selection of a "magic number". E.g. lnglat offset mode's error increases with latitude. Our test datasets are very SF centric.

Comment (georgios-uber) [reply modules/core/src/shaderlib/project/viewport-uniforms.js]: OK, let me try to do that and come up with some numbers.

File: examples/layer-browser/src/examples/core-layers.js [hunk 10]
```diff
@@ -183,6 +183,19 @@ const LineLayerExample = {
...
+const LineLayerExampleNewCoords = {
+  layer: LineLayer,
+  getData: () => dataSamples.routes,
+  props: {
+    id: 'lineLayer',
+    getSourcePosition: d => d.START,
+    getTargetPosition: d => d.END,
+    getColor: d => (d.SERVICE === 'WEEKDAY' ? [255, 64, 0] : [255, 200, 0]),
+    pickable: true,
+    coordinateSystem: COORDINATE_SYSTEM.EXPERIMENTAL_LNG_LAT
```
[the remark this answers is not in the document]
Comment (georgios-uber) [reply examples/layer-browser/src/examples/core-layers.js]: Done

File: modules/core/src/shaderlib/project/project.glsl.js [hunk 11]
```diff
@@ -115,6 +123,15 @@ vec4 project_position(vec4 position) {
...
+  vec4 projected_position = project_position(vec4(position, 1.0), positionLOW);
+  return projected_position.xyz;
+}
+
 vec3 project_position(vec3 position) {
   vec4 projected_position = project_position(vec4(position, 1.0));
```
[the remark this answers is not in the document]
Comment (georgios-uber) [reply modules/core/src/shaderlib/project/project.glsl.js]: OK

File: modules/core/src/shaderlib/project/viewport-uniforms.js (hunk 9 above)
Comment (georgios-uber) [reply modules/core/src/shaderlib/project/viewport-uniforms.js]: Any help on how `offset-accuracy.md` was calculated?

File: docs/upgrade-guide.md [hunk 12]
```diff
@@ -25,6 +25,11 @@ Pixel sizes in line, icon and text layers now match their HTML/SVG counterparts.
 Some previously deprecated `project` module functions have now been removed.
 
 
+#### `is64bitEnabled()` deprecated
```
Comment (ibgreen) [inline docs/upgrade-guide.md]: Nit: Mention class name `Layer.is64...`

File: modules/core/src/shaderlib/project/project.glsl.js [hunk 13]
```diff
@@ -115,13 +124,22 @@ vec4 project_position(vec4 position) {
   return project_offset_(position_modelspace);
 }
 
+vec4 project_position(vec4 position) {
+  return project_position(position, vec2(0.0, 0.0));
```
Comment (ibgreen) [inline modules/core/src/shaderlib/project/project.glsl.js]: Idea: maybe defined a constant `const vec2 zero_64(0.0, 0.0)` and use in all these places to make code more self documenting?

File: modules/core/src/shaderlib/project/viewport-uniforms.js [hunk 14]
```diff
@@ -54,8 +57,20 @@ function calculateMatrixAndOffset({
   let {viewProjectionMatrix} = viewport;
 
   let projectionCenter;
+  let shaderCoordinateSystem = coordinateSystem;
+  let shaderCoordinateOrigin = coordinateOrigin;
+
+  if (coordinateSystem === COORDINATE_SYSTEM.LNGLAT_EXPERIMENTAL) {
+    if (coordinateZoom < LNGLAT_EXPERIMENTAL_ZOOM_THRESHOLD) {
+      shaderCoordinateSystem = COORDINATE_SYSTEM.LNGLAT;
```
Comment (ibgreen) [inline modules/core/src/shaderlib/project/viewport-uniforms.js]: We'll need to define a separate enum for the shader once we make this the default mode. Maybe do that now?

File: modules/core/src/shaderlib/project/viewport-uniforms.js [hunk 15]
```diff
@@ -144,7 +165,13 @@ function calculateViewportUniforms({
   const coordinateZoom = viewport.zoom;
   assert(coordinateZoom >= 0);
 
-  const {projectionCenter, viewProjectionMatrix, cameraPos} = calculateMatrixAndOffset({
+  const {
```
Comment (ibgreen) [inline modules/core/src/shaderlib/project/viewport-uniforms.js]: This code is getting ugly... Too hard to improve?

File: modules/core/src/shaderlib/project/viewport-uniforms.js [hunk 16]
```diff
@@ -181,16 +208,22 @@ function calculateViewportUniforms({
     project_uCameraPosition: cameraPos
   };
 
-  if (coordinateSystem === COORDINATE_SYSTEM.METER_OFFSETS) {
-    const distanceScalesAtOrigin = viewport.getDistanceScales(coordinateOrigin);
+  if (shaderCoordinateSystem === COORDINATE_SYSTEM.METER_OFFSETS) {
```
Comment (ibgreen) [inline modules/core/src/shaderlib/project/viewport-uniforms.js]: Nit: feels like a switch statement would be clearer?

File: docs/upgrade-guide.md (hunk 12 above)
Comment (georgios-uber) [reply docs/upgrade-guide.md]: OK

File: modules/core/src/shaderlib/project/project.glsl.js (hunk 13 above)
Comment (georgios-uber) [reply modules/core/src/shaderlib/project/project.glsl.js]: Done

File: modules/core/src/shaderlib/project/viewport-uniforms.js (hunk 14 above)
Comment (georgios-uber) [reply modules/core/src/shaderlib/project/viewport-uniforms.js]: Next PR.

File: modules/core/src/shaderlib/project/viewport-uniforms.js (hunk 15 above)
Comment (georgios-uber) [reply modules/core/src/shaderlib/project/viewport-uniforms.js]: How?

File: modules/core/src/shaderlib/project/viewport-uniforms.js (hunk 16 above)
Comment (georgios-uber) [reply modules/core/src/shaderlib/project/viewport-uniforms.js]: Next PR.

**Pull request merged**